crates/sol-types/src/types/error.rs

```diff
@@ -405,18 +406,19 @@ impl PanicKind {
     }
 }
 
-/// Returns the revert reason from the given output data. Returns `None` if the
-/// content is not a valid ABI-encoded [`GenericContractError`] or a [UTF-8
-/// string](String) (for Vyper reverts).
-pub fn decode_revert_reason(out: &[u8]) -> Option<String> {
+/// Decodes and retrieves the reason for a revert from the provided output data.
+///
+/// Returns `None` if the content is not a valid ABI-encoded [`GenericContractError`] or
+/// a [UTF-8 string](String) (for Vyper reverts).
+pub fn decode_revert_reason(out: &[u8]) -> Option<GenericRevertReason> {
```

Comment:
should we move this to the type instead:

```rust
impl<T: SolInterface> RevertReason<T> {
  
   fn decode() -> Option<Self>
}
```

Reply:
@mattsse So you mean something like?

```rust
impl<T: SolInterface> RevertReason<T> {
    /// Decodes and retrieves the reason for a revert from the provided output data.
    ///
    /// Returns `None` if the content is not a valid ABI-encoded [`GenericContractError`] or
    /// a [UTF-8 string](String) (for Vyper reverts).
    pub fn decode(out: &[u8]) -> Option<Self> {
        // Try to decode as a generic contract error.
        if let Ok(error) = GenericContractError::abi_decode(out, true) {
            return Some(error.into());
        }

        // If that fails, try to decode as a regular string.
        if let Ok(decoded_string) = core::str::from_utf8(out) {
            return Some(decoded_string.to_string().into());
        }

        // If both attempts fail, return None.
        None
    }
}
```

And then?

```rust
/// Decodes and retrieves the reason for a revert from the provided output data.
///
/// Returns `None` if the content is not a valid ABI-encoded [`GenericContractError`] or
/// a [UTF-8 string](String) (for Vyper reverts).
pub fn decode_revert_reason(out: &[u8]) -> Option<GenericRevertReason> {
    RevertReason::decode(out)
}
```

Reply:
Yeah that makes sense

Reply:
Changes made